sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java

```diff
@@ -17,24 +17,168 @@
  */
 package org.apache.beam.sdk.io.rabbitmq;
 
+import com.google.auto.value.AutoValue;
 import com.rabbitmq.client.AMQP;
 import com.rabbitmq.client.AMQP.BasicProperties;
 import com.rabbitmq.client.Envelope;
 import com.rabbitmq.client.GetResponse;
 import com.rabbitmq.client.LongString;
 import java.io.Serializable;
-import java.util.Arrays;
+import java.time.temporal.ChronoUnit;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Objects;
+import java.util.Optional;
 import javax.annotation.Nullable;
 
 /**
- * It contains the message payload, and additional metadata like routing key or attributes. The main
- * reason of this class is that AMQP.BasicProperties doesn't provide a serializable public API.
+ * Serializable wrapper around rabbitmq's {@link GetResponse}.
+ *
+ * <p>The main purpose of this class is to provide a serializable public API for
+ * AMQP.BasicProperties, which lacks one of its own. Secondarily, introducing the builder pattern
+ * with AutoValue makes this a bit easier to work with.
  */
-public class RabbitMqMessage implements Serializable {
+@AutoValue
```

Comment:
I've rebuilt this using google's AutoValue library which is used liberally throughout this code base. I wanted to be able to modify individual properties (like routing key) easily.